src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs

```diff
@@ -365,22 +365,12 @@ public TransactionNode(Transaction tx)
         /// <param name="connectToNode">The node that will be connected to.</param>
         public static void Connect(CoreNode thisNode, CoreNode connectToNode)
         {
-            var cancellation = new CancellationTokenSource();
-            cancellation.CancelAfter(TimeSpan.FromSeconds(30));
+            if (IsNodeConnectedTo(thisNode, connectToNode))
+                return;
 
-            WaitLoop(() =>
-            {
-                try
-                {
-                    if (IsNodeConnectedTo(thisNode, connectToNode))
-                        return true;
-                    thisNode.CreateRPCClient().AddNode(connectToNode.Endpoint, true);
-                    return true;
```

Comment:
The race condition actually happened because of this line. If we took out the `return` here it would check if the nodes are connected on the next iteration and then exit. But yeah the logic is a bit too much.